### PR TITLE
add rustc_abi debugging attribute

### LINKED ISSUE #115168: Include ABI information in rustc_layout(debug)
It is currently rather hard to figure out which [`PassMode`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/abi/call/enum.PassMode.html) a type will get. It would be great to have some way to display that information, similar to [`rustc_layout(debug)`](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7abc008aed2669466d4ebe79ee7767fe). The logic for computing `PassMode` is currently here and we'd have to find some way to invoke it:

https://github.com/rust-lang/rust/blob/795ade084ac18e2198faeb63a783d2b8e2108c42/compiler/rustc_ty_utils/src/abi.rs#L404-L416

@eddyb @bjorn3 any advice on how to implement this? Or is it meaningless to ask about the `PassMode` of a type, since this has to always be considered in the context of a particular function signature?

### PR DESCRIPTION
**Pull request by RalfJung**
This is the call ABI equivalent of `rustc_layout(debug)`.

Fixes https://github.com/rust-lang/rust/issues/115168
r? @bjorn3

### CODE REVIEW

File: compiler/rustc_passes/src/abi_test.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+                    name => {
+                        tcx.sess.emit_err(UnrecognizedField { span: meta_item.span(), name });
+                    }
+                }
+            }
+        }
+
+        Err(FnAbiError::Layout(layout_error)) => {
+            tcx.sess.emit_fatal(Spanned {
```
Comment (compiler-errors) [inline compiler/rustc_passes/src/abi_test.rs:77]: Why fatal? I would just error with a message saying that the signature is too polymorphic or something (at least if it's a `LayoutError::Unknown`).

File: compiler/rustc_passes/src/abi_test.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            // Not sure what to do here, but `LayoutError::Unknown` seems reasonable?
+            let ty = tcx.type_of(item_def_id).instantiate_identity();
+            tcx.sess.emit_fatal(Spanned {
+                node: LayoutError::Unknown(ty).into_diagnostic(),
+
+                span: tcx.def_span(item_def_id),
+            });
+        }
+        Err(_guaranteed) => return,
+    };
```
Comment (compiler-errors) [inline compiler/rustc_passes/src/abi_test.rs:52]: This should always resolve to `Instance::mono(item_def_id, args)` -- not sure if you need to actually resolve?

File: compiler/rustc_passes/src/abi_test.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_passes/src/abi_test.rs:77]: Oh, I see. FnAbiError only implements IntoDiagnostic<!> :upside_down_face:

**Review by compiler-errors — APPROVED** [review]
r=me ideally changing that `Instance::resolve` call to a call to `Instance::mono`, unless you want a specific review from bjorn3

File: tests/ui/abi/debug.stderr [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+LL | fn test(_x: u8) -> bool { true }
+   | ^^^^^^^^^^^^^^^^^^^^^^^
+
+error: fn_abi_of_instance(assoc_test) = FnAbi {
+           args: [
+               ArgAbi {
+                   layout: TyAndLayout {
+                       ty: &S,
+                       layout: Layout {
+                           size: Size(8 bytes),
```
Comment (compiler-errors) [inline tests/ui/abi/debug.stderr:101]: Gotta normalize size of ptrs too I guess :smile:

File: compiler/rustc_passes/src/abi_test.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_passes/src/abi_test.rs:77]: This is LayoutError. Also this matches what `layout_test` does. :shrug: 

FnAbiError doesn't implement any kind of IntoDiagnstic that I could find.

File: compiler/rustc_passes/src/abi_test.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_passes/src/abi_test.rs:52]: Then it will `bug!` on a generic item though, won't it?

Comment (compiler-errors) [reply compiler/rustc_passes/src/abi_test.rs:53]: Ah, yeah, I guess it's designed to prevent you from side-stepping resolution like this.

Comment (RalfJung) [reply compiler/rustc_passes/src/abi_test.rs:53]: The neat thing is that if the generics do not affect the ABI, `rustc_abi(debug)` actually works on generic items. :)  I just added a test for that.

Comment (compiler-errors) [reply compiler/rustc_passes/src/abi_test.rs:53]: Yeah, which is why I had suggested mono earlier, I thought it just constructed an InstanceDef::Item..

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**